dynamic_programming/longest_common_string.cpp

```diff
@@ -1,53 +1,148 @@
-#include <iosrteam>
-using namespace std;
+/**
+ * @file
+ * @brief contains the definition of the function longest_common_string_length
+ * @details
+ * the function longest_common_string_length computes the length
+ * of the longest common string which can be created of two input strings
+ * by removing characters from them
+ *
+ * @author [Nikhil Arora](https://github.com/nikhilarora068)
+ * @author [Piotr Idzik](https://github.com/vil02)
+ */
 
-int max(int a, int b) { return (a > b) ? a : b; }
+#include <cassert>   /// for assert
+#include <iostream>  /// for std::cout
+#include <string>    /// for std::string
+#include <utility>   /// for std::move
+#include <vector>    /// for std::vector
 
-int main() {
-    char str1[] = "DEFBCD";
-    char str2[] = "ABDEFJ";
-    int i, j, k;
-    int n = strlen(str1) + 1;
-    int m = strlen(str2) + 1;
-    // cout<<n<<" "<<m<<"\n";
-    int a[m][n];
-
-    for (i = 0; i < m; i++) {
-        for (j = 0; j < n; j++) {
-            if (i == 0 || j == 0)
-                a[i][j] = 0;
-
-            else if (str1[i - 1] == str2[j - 1])
-                a[i][j] = a[i - 1][j - 1] + 1;
-
-            else
-                a[i][j] = 0;
-        }
-    }
+/**
+ * @brief computes the length of the longest common string created from input
+ * strings
+ * @details has O(str_a.size()*str_b.size()) time and memory complexity
+ * @param string_a first input string
+ * @param string_b second input string
+ * @returns the length of the longest common string which can be strated from
+ * str_a and str_b
+ */
+std::size_t longest_common_string_length(const std::string& string_a,
+                                         const std::string& string_b) {
+    const auto size_a = string_a.size();
+    const auto size_b = string_b.size();
+    std::vector<std::vector<std::size_t>> sub_sols(
+        size_a + 1, std::vector<std::size_t>(size_b + 1, 0));
 
-    /*for(i=0;i<m;i++)
-        {
-            for(j=0;j<n;j++)
-                cout<<a[i][j]<<" ";
-            cout<<"\n";
-        }*/
-
-    int ma = -1;
-    int indi, indj;
-    for (i = 0; i < m; i++) {
-        for (j = 0; j < n; j++) {
-            if (a[i][j] > ma) {
-                ma = a[i][j];
-                indi = i;
-                indj = j;
+    const auto limit = static_cast<std::size_t>(-1);
+    for (std::size_t pos_a = size_a - 1; pos_a != limit; --pos_a) {
+        for (std::size_t pos_b = size_b - 1; pos_b != limit; --pos_b) {
+            if (string_a[pos_a] == string_b[pos_b]) {
+                sub_sols[pos_a][pos_b] = 1 + sub_sols[pos_a + 1][pos_b + 1];
+            } else {
+                sub_sols[pos_a][pos_b] = std::max(sub_sols[pos_a + 1][pos_b],
+                                                  sub_sols[pos_a][pos_b + 1]);
             }
         }
     }
 
-    cout << str1 << "\n";
-    cout << str2 << "\n";
+    return sub_sols[0][0];
+}
+
+struct TestCase {
```

Comment:
structs must be documented

Reply:
No, they do not have to be.

Reply:
They do. The contributing guidelines says so. You must document structs.

Reply:
They must be documented in order for Doxygen to register it and **all of its content**.

Reply:
> They must be documented for Doxygen to register it and **all of its contents**.

What do you mean by that? When I run Doxygen, the page for `TestCase` looks like this:
![TestCasedoc](https://github.com/TheAlgorithms/C-Plus-Plus/assets/65706193/567fb52e-6733-4e06-bb8f-c0b1c14e3afb)

Reply:
which isnt enough see how there is no information about what anything is?

Reply:
ill fix this in another pull request im resolving for now

Reply:
This is generated thanks to documenting the function.
https://thealgorithms.github.io/C-Plus-Plus/d1/d21/quick__sort_8cpp.html#a483bb8ccf42aaf7375a83e91490eda1e

![image](https://github.com/TheAlgorithms/C-Plus-Plus/assets/51391473/d0e23121-3ef4-42e4-85a3-27f54bd758f9)

Reply:
Behavior of doxygen on functions, which do not have _documentation strings_ can be configured. What could be a reason that current configuration does not generate call graphs for functions not having documentation strings? I think it comes from the fact that a function _trivial enough_ should not create an extra _noise_ in the documentation (in other words: current configuration allows the programmer to decide which parts of code are really important and which not).

By the way, in the example which you have provided, the call graph is incorrect. The `tests` function also calls:
- `sorting::quick_sort::quick_sort` (that is what we test there, right?),
- `sorting::quick_sort::show`,
- some constructors of `std::vector`.

Also the word `normal` [here](https://github.com/TheAlgorithms/C-Plus-Plus/blame/2fd530cba1a798fe2f9bdfe835c8179072de2394/sorting/quick_sort.cpp#L136) could be replaced by something different (cf. [Normal number](https://en.wikipedia.org/wiki/Normal_number)).

Reply:
> By the way, in the example which you have provided, the call graph is incorrect.

That graph is being generated by Doxygen, weird that it is showing up like that.
I understand these functions should not be documented, but this still helps in Doxygen fully registering them, so, please add documentation to those.